File: framework/references/examples/node-backend/repository-example.ts
import { Result } from '../common/result';
import { User } from '../domain/user';
import { AppDataSource } from '../config/database';

export interface IUserRepository {
  save(user: User): Promise<Result<User>>;
  findById(id: string): Promise<Result<User>>;
}

export class UserRepository implements IUserRepository {
  async save(user: User): Promise<Result<User>> {
    try {
      await AppDataSource.getRepository(User).save(user);
      return Result.ok(user);
    } catch (err: any) {
      if (err.code === '23505') {
        return Result.fail('Email already exists');
      }
      return Result.fail(`Failed to save user: ${err.message}`);
    }
  }

  async findById(id: string): Promise<Result<User>> {
    const user = await AppDataSource.getRepository(User).findOneBy({ id });
    if (!user) return Result.fail('User not found');
    return Result.ok(user);
  }
}

File: framework/references/examples/backend-dotnet/repository-example.cs
using Microsoft.EntityFrameworkCore;
using YourProject.Domain;
using YourProject.Common;

namespace YourProject.Infrastructure.Repositories;

public interface IUserRepository
{
    Task<Result<User>> SaveAsync(User user, CancellationToken ct);
    Task<Result<User>> FindByIdAsync(string id, CancellationToken ct);
}

public sealed class UserRepository : IUserRepository
{
    private readonly AppDbContext _db;

    public UserRepository(AppDbContext db)
    {
        _db = db;
    }

    public async Task<Result<User>> SaveAsync(User user, CancellationToken ct)
    {
        try
        {
            _db.Users.Add(user);
            await _db.SaveChangesAsync(ct);
            return Result<User>.Success(user);
        }
        catch (DbUpdateException ex) when (ex.InnerException?.Message.Contains("duplicate") == true)
        {
            return Result<User>.Failure("Email already exists");
        }
    }

    public async Task<Result<User>> FindByIdAsync(string id, CancellationToken ct)
    {
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == id, ct);
        return user is not null
            ? Result<User>.Success(user)
            : Result<User>.Failure("User not found");
    }
}
